src/Parser/Parser.hs

```diff
@@ -360,35 +364,50 @@ xtype = (try $ do
 -- | Parse just a type, before it is possibly extended
 xtype' :: Parser Xtype
 xtype' =
-  (try $ capIdentifier >>= lookupSyn)
+  (try typeName)
   <|>
   (try $ X <$> (pure Top) <*> enum) -- Plain enum
   <|>
   (try $ X <$> btype <*> (pure S.empty))
   <|>
   (try $ Tup <$> parens (lexeme ((:) <$> (xtype <* comma) <*> (commaSep1 xtype))))
 
+-- | Parse an xtype, but only if it is one of the base types or it has been declared
+--   and named (with the type keyword)
+namedType :: Parser Xtype
+namedType =
+   (try $ notExtended typeName)
+   <|>
+   (try $ Tup <$> parens (lexeme ((:) <$> (namedType <* comma) <*> (commaSep1 namedType))))
+   <|>
+   (notExtended $ X <$> btype <*> pure S.empty)
+   <?>
+   "a pre-existing or previously-declared type"
+      where
+         notExtended :: Parser a -> Parser a
+         notExtended n = n <* (notFollowedBy (string "&"))
+
 -- | Function types
 ftype :: Parser Ftype
 ftype = do
-  x <- xtype
+  x <- namedType
   reservedOp "->"
-  r <- xtype
+  r <- namedType
   return $ Ft x r
 
--- | Types
+-- | Types that appear in signatures
 typ :: Parser Type
 typ =
   (try $ (do
+            _p <- namedType <* notFollowedBy (reservedOp "->")
```

Comment:
Not sure about flipping the order of parsing a type for a function vs. a value eq. Looks like it'll work with the reservedOp "->" lookahead, but it's similar to pattern matching in haskell in general, with the most general case usually coming last. Unless there was a specific issue with having to flip the order that I'm missing.

Reply:
It was intentional. The first example is an error message with the parser as-is (veq then feq). The second is with the original order (feq then veq).

![good](https://user-images.githubusercontent.com/6590681/97749144-c8787780-1aab-11eb-958e-34a00e19000d.png)
![bad](https://user-images.githubusercontent.com/6590681/97749149-cadad180-1aab-11eb-898d-08fa9f8a0bae.png)

In the second example: 
1. an attempt to parse a feq signature fails
2. an attempt to parse a veq signature fails because `->` should not appear in a veq signature
    - it reports this as the source of the parse error
    - technically, this is correct source of the parse error, however, it is very misleading
    - the root cause is the `&` in the signature, so I swapped the order to ensure it is reported as such

If you have a better way to achieve it, I would happily welcome it.


Note: when I swapped the order to produce the image of my bad example, I also changed `notfollowedBy (reservedOp "->")` to `notfollowedBy (string "->")` to make it clear that the `->` was causing the failure because reservedOp parses a unit value, i.e. `()` and then the error would read `unexpected ()`.

Reply:
Ahhhhhhhhh. Okay, yeah I definitely like the new ordering then if it produces an improved message, that is much more effective at conveying the type error than the prior; good choice 👍 . I can't think of a better way to do that, this is good.